src/Microsoft.OData.Core/Evaluation/ODataConventionalEntityMetadataBuilder.cs

```diff
@@ -408,7 +408,11 @@ private Uri ComputeEditLink()
             Debug.Assert(this.ResourceMetadataContext != null && this.ResourceMetadataContext.TypeContext != null, "this.resourceMetadataContext != null && this.resourceMetadataContext.TypeContext != null");
             if (this.ResourceMetadataContext.ActualResourceTypeName != this.ResourceMetadataContext.TypeContext.NavigationSourceEntityTypeName)
             {
-                uri = this.UriBuilder.AppendTypeSegment(uri, this.ResourceMetadataContext.ActualResourceTypeName);
+                // In some scenarios, the resource Id will already contain the cast segment
+                if (!uri.ToString().EndsWith(string.Concat('/', this.ResourceMetadataContext.ActualResourceTypeName), StringComparison.Ordinal))

```

Comment:
`string.Concat` does not have an overload that accepts a `char` as the first argument, so this method call will probably invoke an overload that takes `object` as an argument, which means the `/` char will be boxed and cause an allocation.

I suspect it would be more efficient to call `string.Concat("/", ...)` instead.

Here's an illustration of the IL generated for both variants: https://sharplab.io/#v2:C4LglgNgNAJiDUAfAAgJgIwFgBQyDMABGgQMIEDeOB1Rhy6ADKQPYB2AxgIbAkAWnAJwAU9JgGcAlBSo1ZyAOwEAysAFhWAcwB0JNl2BCA5AHpDUApIDcM6gF8bBB/iKMWHbirWaRrydOyycoqiOnrcQgBExhHmVg722LZAA

Reply:
Also, it may be possible to match the string ending without performing a concatenation, but maybe the code complexity this would introduce is not worth it for this edge case (and it's not guaranteed that it would perform better). e.g.,: `if (foo.EndsWith(bar) && foo[foo.length - bar.length - 1] == '/' ...)` (there's probably an off-by-one error in there, but you get the idea).

Reply:
This logic seems to hinge on the fact `ActualResourceTypeName` is a fully-qualified name, and in this context only a type segment would be a fully qualified name. Is that correct?